### ISSUE TITLE
Feature Request: Implement Chaining for Async Matchers

### ISSUE THREAD

**Issue by schaable** [NomicFoundation/hardhat#4235]
### Describe the feature

The hardhat-chai-matchers plugin currently lacks support for chaining with async matchers such as `reverted`, `revertedWith`, `changeEtherBalance`, etc.

Chaining would enable users to create more expressive and readable assertions by combining multiple matchers, improving the testing experience.

Example of the desired chaining:

```js
await expect(contract.method(...))
  .to.be.revertedWithCustomError(...)
  .and.to.not.changeTokenBalance(...);
```

As a temporary workaround, you can create separate assertions for each async matcher:

```js
const tx = contract.method(...);
await expect(tx).to.be.revertedWithCustomError(...);
await expect(tx).to.not.changeTokenBalance(...);
```

If you believe implementing chaining for async matchers would be valuable and beneficial for your testing workflows, please consider upvoting this issue to show your interest and support.

**Comment by sherpya** [comment]
I always known there was something strange in my tests :)

**Comment by fvictorio** [comment]
Yes, `.emit` is the only one that can be chained (and only to itself).

@schaable we should update the plugin's readme to add `emit` to the list of matchers that do not support chaining, and add a clarification saying "The only exception is chaining multiple .emit matchers" or something like that.

**Comment by nikicat** [comment]
Hello, how many upvotes are required for this feature to be considered important enough for implementation?

**Comment by fvictorio** [comment]
Hi @nikicat. We use upvotes to gauge interest, but there isn't a linear relationship between them and the likelihood of us working on an issue. There are other considerations, like how hard it is to implement, whether it enables new use cases or is a quality of life improvement, and more.

In this case, this is much harder to implement than it seems. Vanilla Chai in general is not easy to extend, and _async_ chai is just hell. And while we understand it would be more convenient to be able to chain assertions, the thing is that it's not that hard to work around the lack of this feature.

In addition to that, viem now exists as a strong alternative to ethers, which means that we need to support two ways to write assertions, leaving us even less bandwidth for nice-to-have improvements like this.

I don't want to downplay this feature request. There are clearly many people that want it, and I can relate to how ugly it is to work around this by assigning a transaction to a variable to assert multiple things about it. I hope we eventually can implement this. But the reality today is that it won't happen any time soon.
